apps/internal/base/internal/storage/storage.go

```diff
@@ -262,9 +256,10 @@ func (m *Manager) aadMetadata(ctx context.Context, authorityInfo authority.Info)
 }
 
 func (m *Manager) readAccessToken(homeID string, envAliases []string, realm, clientID string, scopes []string) (AccessToken, error) {
-	cache := m.Contract()
-
-	for _, at := range cache.AccessTokens {
+	m.contractMu.RLock()
+	defer m.contractMu.RUnlock()
+	// TODO: linear search (over a map no less) is slow for a large number of tokens
```

Comment:
I'm not sure if the TODO is valid??  It's all going to depend on the average case.  If the average is small < 100 tokens or this isn't in a n^2 scenario, I doubt this will matter.

However, if our case doesn't resemble that, then we are either going to have to:
- write multiple maps keyed on HomeAccountID, Realm and ClientID and do a union of 3 hash lookups
- store another map that has a combo key just for this "HAI:Realm:CID"
- have a different data structure that works for this case

These are all actually expensive.  Your either making multiple maps or are dealing with a data structure that is:
A) Hidden 
or 
B) Serializes into key/values when using JSON.  

With this said, is this valid TODO, is this something we note why we are iterating over a map, .......??????

Reply:
After I made this change I was surprised to see that token retrieval wasn't O(1) like I expected, so I profiled it.  It's the dominating node in the profile for a large number of tokens.

I agree that it likely doesn't matter for real-world scenarios, however I wanted to ensure this behavior is documented in case it becomes a problem in the future.

Reply:
Do me a favor, add the text of the above to give some explanation fo why this isn't done then.  This lets someone in the future understand the kind of things they need to do if this becomes a problem.

Reply:
Done